### PR TITLE
Rework ImplicitContext

### PR DESCRIPTION
**Pull request by pepone**
This PR simplifies the implementation of the implicit context. While it may not be as optimized as before for the per-thread case, it is much simpler.

### CODE REVIEW

File: cpp/src/Ice/Instance.h [hunk 1]
```diff
@@ -196,6 +193,16 @@ class Instance : public std::enable_shared_from_this<Instance>
     std::mutex _setBufSizeWarnMutex;
     mutable std::recursive_mutex _mutex;
     std::condition_variable_any _conditionVariable;
+
+    enum ImplicitContextKind
```
[the remark this answers is not in the document]
Comment (pepone) [reply cpp/src/Ice/Instance.h:197]: Not clear why you mean, we cannot declare a thread_local at class scope.

File: cpp/src/Ice/Instance.cpp [hunk 2]
```diff
@@ -1283,6 +1304,29 @@ IceInternal::Instance::initialize(const Ice::CommunicatorPtr& communicator)
     }
 }
 
+const Ice::ImplicitContextIPtr&
+IceInternal::Instance::getImplicitContext() const
+{
+    if(_implicitContextKind == ImplicitContextKind::PerThread)
+    {
+        lock_guard lock(perThreadImplicitContextMutex);
```
[the remark this answers is not in the document]
Comment (pepone) [reply cpp/src/Ice/Instance.cpp:1312]: You are right it is not required.

File: cpp/src/Ice/Instance.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (pepone) [reply cpp/src/Ice/Instance.h:197]: Here `thread_local ImplicitContextPtr perThreadImplicitContext;` is shared by all instances, which is not what we want

[the remark this answers is not in the document]
Comment (bernardnormier) [reply cpp/src/Ice/Instance.h:197]: We want to keep the same logic: all these context instances are per-communicator. And sometimes also per-thread.

File: cpp/src/Ice/ImplicitContext.cpp [hunk 3]
```diff
@@ -2,23 +2,112 @@
...
 
-#ifndef ICE_API_EXPORTS
-#   define ICE_API_EXPORTS
-#endif
 #include <Ice/ImplicitContext.h>
-#include <IceUtil/PushDisableWarnings.h>
-#include <Ice/InputStream.h>
 #include <Ice/OutputStream.h>
-#include <IceUtil/PopDisableWarnings.h>
+#include <Ice/Object.h>
```
Comment (bernardnormier) [inline cpp/src/Ice/ImplicitContext.cpp:7]: Not clear why we need Object.h here.

File: cpp/src/Ice/ImplicitContext.cpp [hunk 4]
```diff
@@ -2,23 +2,112 @@
...
+    {
+        Context combined = proxyCtx;
+        combined.insert(_context.begin(), _context.end());
+        lock.unlock();
+        s->write(combined);
+    }
+}
+
+void
+ImplicitContext::combine(const Context& proxyCtx, Context& ctx) const
```
Comment (bernardnormier) [inline cpp/src/Ice/ImplicitContext.cpp:97]: You should keep the same parameter names are in the declaration / doc-comments.

File: cpp/src/Ice/Instance.cpp [hunk 5]
```diff
@@ -1283,6 +1301,33 @@ IceInternal::Instance::initialize(const Ice::CommunicatorPtr& communicator)
...
+            }
+            else
+            {
+                return it->second;
+            }
+        }
+
+        default:
+        {
+            return _sharedImplicitContext;
```
Comment (bernardnormier) [inline cpp/src/Ice/Instance.cpp:1326]: When ImplicitContextKind::None, we return a dereferenced null shared_ptr?

**Review by bernardnormier — COMMENTED** [review]
Looks like it's still WIP.

File: cpp/src/Ice/ImplicitContext.cpp (hunk 3 above)
Comment (pepone) [reply cpp/src/Ice/ImplicitContext.cpp:7]: We don't, removed

File: cpp/src/Ice/Instance.cpp [hunk 6]
```diff
@@ -1283,6 +1301,33 @@ IceInternal::Instance::initialize(const Ice::CommunicatorPtr& communicator)
...
+            auto it = perThreadImplicitContextMap.find(this);
+            if (it == perThreadImplicitContextMap.end())
+            {
+                auto r = perThreadImplicitContextMap.emplace(make_pair(this, std::make_shared<ImplicitContext>()));
+                return r.first->second;
+            }
+            else
+            {
+                return it->second;
+            }
```
Comment (pepone) [inline cpp/src/Ice/Instance.cpp:1321]: This code replaces the old `PerThreadImplicitContext` we keep a per thread, per instance ImplicitContext.

File: cpp/src/Ice/Instance.cpp (hunk 5 above)
Comment (bernardnormier) [reply cpp/src/Ice/Instance.cpp:1326]: I assume the caller ensures this never happens, so I would add an assert to make this clearer:
```cpp
case ImplicitContextKind::PerThread:
  ...
case ImplicitContextKind::Shared:
 ...
default:
   assert(0); // never happens
```

Comment (pepone) [reply cpp/src/Ice/Instance.cpp:1326]: The caller OutgoingAsync checks for null:

https://github.com/zeroc-ice/ice/blob/bf991f4f7604869517f9211f18a6048239521551/cpp/src/Ice/OutgoingAsync.cpp#L696-L705

The null return is documented here:

https://github.com/zeroc-ice/ice/blob/bf991f4f7604869517f9211f18a6048239521551/cpp/include/Ice/Communicator.h#L231-L236

**Review by bernardnormier — APPROVED** [review]
Looks good.

**Review by externl — APPROVED** [review]
LGTM!

**Pull request merged**